Static files overhaul (#35967)
dotnet/AspNetCore.Docs: aspnetcore/fundamentals/static-files.md

---
title: Static files in ASP.NET Core
author: wadepickett
description: Learn how to serve and secure static files and configure Map Static Assets endpoint conventions and Static Files Middleware in ASP.NET Core web apps.
monikerRange: '>= aspnetcore-3.1'
ms.author: wpickett
ms.custom: mvc
ms.date: 08/20/2025
uid: fundamentals/static-files
---
# Static files in ASP.NET Core

[!INCLUDE[](~/includes/not-latest-version.md)]

Static files, also called static assets, are files an ASP.NET Core app that aren't dynamically generated and served directly to clients on request, such as HTML, CSS, image, and JavaScript files.

For Blazor static files guidance, which adds to or supersedes the guidance in this article, see <xref:blazor/fundamentals/static-files>.

:::moniker range=">= aspnetcore-9.0"

> ## Map Static Assets (`MapStaticAssets`)

Reviewer: Can we start with some basics?:

- To enable static file handling in ASP.NET Core, call `MapStaticAssets`
- You put any files you want to be served as static files in your wwwroot folder (also called the web root)
- By default, only files in the wwwroot folder will be addressable, so you don't need to worry about the rest of your code.
- Only files with specific file extensions mapped to supported media types will be treated as static web assets.
- Your static web assets will get discovered at build time and optimized using content-based finger printing and precompession.
- At runtime, the discovered static web assets will be exposed as endpoints and will have appropriate HTTP headers applied, like caching headers, content type, etc.
- Note that you can also serve static web assets from referenced projects and packages. Link to related content.

<xref:Microsoft.AspNetCore.Builder.StaticAssetsEndpointRouteBuilderExtensions.MapStaticAssets%2A> maps static files produced during the build as endpoints with the following features:

* An asset is served once until the file changes or the browser clears its cache. The [`ETag`](https://developer.mozilla.org/docs/Web/HTTP/Headers/ETag) and [Last-Modified](https://developer.mozilla.org/docs/Web/HTTP/Headers/Last-Modified) headers are set.
* The browser is prevented from using stale assets after an app is updated. The [`Last-Modified`](https://developer.mozilla.org/docs/Web/HTTP/Headers/Last-Modified) header is set.
* The appropriate [caching headers](https://developer.mozilla.org/docs/Web/HTTP/Headers/Cache-Control) are set on the response.

> * [Caching Middleware](xref:performance/caching/middleware) is used.

Reviewer: The Caching middleware isn't actually a feature of MapStaticAssets. It's a separate independent feature.

```suggestion
```

* [Compressed](/aspnet/core/performance/response-compression) versions of the assets are served when possible. This optimization doesn't include minification.
* A [CDN](/microsoft-365/enterprise/content-delivery-networks?view=o365-worldwide&preserve-view=true) can be used to serve the assets closer to the user.
* Assets are [fingerprinted](https://wikipedia.org/wiki/Fingerprint_(computing)) to prevent reuse of old files.

<xref:Microsoft.AspNetCore.Builder.StaticAssetsEndpointRouteBuilderExtensions.MapStaticAssets%2A>:

* Integrates the information gathered about static web assets during the build and publish process with a runtime library that processes this information to optimize file serving to the browser.
* Are routing endpoint conventions that optimize the delivery of static assets in an app. It's designed to work with all UI frameworks, including Blazor, Razor Pages, and MVC.

> ## Static Files Middleware (`UseStaticFiles`)

Reviewer: I'd like to avoid presenting the reader with a choice of how to handle static files right at the start of this article. Instead, I think the article should start off focused on basic static file handling using MapStaticAssets, almost as if we were documenting static file handling in ASP.NET Core for the first time and MapStaticAssets is the only thing available. After we've covered all the basic stuff, we can then cover scenarios that aren't supported by MapStaticAssets and that require the static file middleware.

## `MapStaticAssets` versus `UseStaticFiles`

<xref:Microsoft.AspNetCore.Builder.StaticAssetsEndpointRouteBuilderExtensions.MapStaticAssets%2A> is available in ASP.NET Core in .NET 9 or later. <xref:Microsoft.AspNetCore.Builder.StaticFileExtensions.UseStaticFiles%2A> must be used in versions prior to .NET 9.

<xref:Microsoft.AspNetCore.Builder.StaticFileExtensions.UseStaticFiles%2A> serves static files, but it doesn't provide the same level of optimization as <xref:Microsoft.AspNetCore.Builder.StaticAssetsEndpointRouteBuilderExtensions.MapStaticAssets%2A>. <xref:Microsoft.AspNetCore.Builder.StaticAssetsEndpointRouteBuilderExtensions.MapStaticAssets%2A> is optimized for serving assets that the app has knowledge of at runtime. If the app serves assets from other locations, such as disk or embedded resources, <xref:Microsoft.AspNetCore.Builder.StaticFileExtensions.UseStaticFiles%2A> should be used.

> Map Static Assets provides the following benefits that aren't available when calling <xref:Microsoft.AspNetCore.Builder.StaticFileExtensions.UseStaticFiles%2A>:

Reviewer: Another benefit of MapStaticAssets is that it is based on endpoint routing, so it works with other endpoint aware features like authorization.

```

:::moniker range=">= aspnetcore-9.0"

## Short-circuit the middleware pipeline

To avoid running the entire middleware pipeline after a static asset is served, which matches the behavior of <xref:Microsoft.AspNetCore.Builder.StaticFileExtensions.UseStaticFiles%2A>, call <xref:Microsoft.AspNetCore.Builder.RouteShortCircuitEndpointConventionBuilderExtensions.ShortCircuit%2A> on <xref:Microsoft.AspNetCore.Builder.StaticAssetsEndpointRouteBuilderExtensions.MapStaticAssets%2A>. Calling <xref:Microsoft.AspNetCore.Builder.RouteShortCircuitEndpointConventionBuilderExtensions.ShortCircuit%2A> immediately executes the endpoint and returns the response, preventing other middleware from executing for static asset requests:

```csharp
app.MapStaticAssets().ShortCircuit();
```

## Control static file caching

During development, the framework overrides cache headers to prevent browsers from caching static files. This helps ensure that the latest version of files are used when files change, avoiding issues with stale content. In production, the correct cache headers are set, allowing browsers to cache static assets as expected.

> To disable this behavior, set `EnableStaticAssetsDevelopmentCaching` to `false` in the Development environment's app setting file (`appsettings.Development.json`).

Reviewer: I think this property might also be specific to MapStaticAssets, but I'm not 100% sure.

By default, for requests to `/`:

* In the development environment, `wwwroot/Index.html` is returned.
* In any environment other than development, `wwwroot-custom/Index.html` is returned.

To ensure assets from `wwwroot-custom` are always returned, use ***one*** of the following approaches:

* Delete duplicate-named assets in `wwwroot`.

* Set `ASPNETCORE_ENVIRONMENT` in `Properties/launchSettings.json` to any value other than `Development`.

* Disable static web assets by setting `<StaticWebAssetsEnabled>` to `false` in the app's project file. ***WARNING:*** Disabling static web assets disables [Razor class libraries](xref:razor-pages/ui-class).

* Add the following XML to the project file:

  ```xml
  <ItemGroup>
    <Content Remove="wwwroot\**" />
  </ItemGroup>
  ```

The following code updates <xref:Microsoft.AspNetCore.Hosting.IWebHostEnvironment.WebRootPath%2A> to a non-Development value (`Staging`), guaranteeing duplicate content is returned from `wwwroot-custom` rather than `wwwroot`:

```csharp
var builder = WebApplication.CreateBuilder(new WebApplicationOptions
{
    Args = args,
    EnvironmentName = Environments.Staging,
    WebRootPath = "wwwroot-custom"
});
```

> When developing a server-side Blazor app and testing locally, see <xref:blazor/fundamentals/static-files#static-files-in-non-development-environments>.

Author: There were behavioral deltas when the article was set up. As a general principal, only deltas (or Blazor-focused scenarios) were covered by the Blazor-specific articles. I'll double-check and get back to you, but it might be Tuesday, given that I'm trying to get the *Environments* article worked wrapped up before the holiday weekend.

```

In the request processing pipeline after the existing call to either <xref:Microsoft.AspNetCore.Builder.StaticAssetsEndpointRouteBuilderExtensions.MapStaticAssets%2A> (.NET 9 or later) or <xref:Microsoft.AspNetCore.Builder.StaticFileExtensions.UseStaticFiles%2A> (.NET 8 or earlier):

```csharp
app.UseStaticFiles(new StaticFileOptions
{
    FileProvider = new PhysicalFileProvider(
        Path.Combine(builder.Environment.ContentRootPath, "ExtraStaticFiles")),
    RequestPath = "/static-files"
});
```

In the preceding code, the `ExtraStaticFiles` directory hierarchy is exposed publicly via the `StaticFiles` URL segment. A request to `https://{HOST}/StaticFiles/images/red-rose.jpg`, where the `{HOST}` placeholder is the host, serves the `red-rose.jpg` file.

The following markup references `ExtraStaticFiles/images/red-rose.jpg`:

```html
<img src="static-files/images/red-rose.jpg" alt="A red rose" />
```

For the preceding example, tilde-slash notation is supported in Razor Pages and MVC views (`src="~/StaticFiles/images/red-rose.jpg"`), not for Razor components in Blazor apps.

## Serve files from multiple locations

:::moniker range=">= aspnetcore-6.0"

> *The guidance in this section applies to Razor Pages and MVC apps. For guidance that applies to Blazor Web Apps, see <xref:blazor/fundamentals/static-files#serve-files-from-multiple-locations>.*

Reviewer: Isn't the guidance the same? Can we remove this duplication?

Author: That's what I meant about deltas. It wasn't the same IIRC, but I'll have to look. I'll check and get back to you ... probably Tuesday morning.

In the following example, the exclusion patten adds the static files in the `lib/icons` folder, which represents a hypothetical batch of icons:

```xml
<StaticWebAssetEndpointExclusionPattern>
  $(StaticWebAssetEndpointExclusionPattern);lib/icons/**
</StaticWebAssetEndpointExclusionPattern>
```

After HTTPS Redirection Middleware (`app.UseHttpsRedirection();`) processing in the `Program` file:

* Call <xref:Microsoft.AspNetCore.Builder.StaticFileExtensions.UseStaticFiles%2A> to handle the excluded files (`lib/icons/**`) and any other files not covered by <xref:Microsoft.AspNetCore.Builder.StaticAssetsEndpointRouteBuilderExtensions.MapStaticAssets%2A>.
* Call <xref:Microsoft.AspNetCore.Builder.StaticAssetsEndpointRouteBuilderExtensions.MapStaticAssets%2A> after <xref:Microsoft.AspNetCore.Builder.StaticFileExtensions.UseStaticFiles%2A> to handle critical application files (CSS, JS, images).

```csharp
app.UseStaticFiles();

app.UseAuthorization();

app.MapStaticAssets();
```

> ## Static file authorization

Reviewer: I'm a bit confused about what this section is trying to communicate. It sounds like it's trying to explain how you can add authorization checks for your static files. But then the first thing it shows is enabling anonymous access to static web assets. Shouldn't this section start off with showing that static web assets can have authorization policies applied just like any other endpoint?

The rest of the section seems to be about workarounds for enabling authorization when using the static files middleware. This content is certainly important for folks still using the static files middleware, but it dominates this section in a way that I don't think we want.

monikerRange: '>= aspnetcore-3.1'
ms.author: wpickett
ms.custom: mvc
ms.date: 09/02/2025
uid: fundamentals/static-files
---
# Static files in ASP.NET Core

[!INCLUDE[](~/includes/not-latest-version.md)]

Static files, also called static assets, are files an ASP.NET Core app that aren't dynamically generated and served directly to clients on request, such as HTML, CSS, image, and JavaScript files.

For Blazor static files guidance, which adds to or supersedes the guidance in this article, see <xref:blazor/fundamentals/static-files>.

:::moniker range=">= aspnetcore-9.0"

To enable static file handling in ASP.NET Core, call <xref:Microsoft.AspNetCore.Builder.StaticAssetsEndpointRouteBuilderExtensions.MapStaticAssets%2A>. 

By default, static files are stored within the project's [web root](xref:fundamentals/index#web-root) directory. The default directory is `{CONTENT ROOT}/wwwroot`, where the `{CONTENT ROOT}` placeholder is the app's [content root](xref:fundamentals/index#content-root). Only files in the wwwroot folder will be addressable, so you don't need to worry about the rest of your code.

> Only files with specific file extensions mapped to supported media types are treated as static web assets.

Reviewer: @javiercn Is there more we should say about the default media type mapping for static web assets and how it can be configured?

Reviewer 2: Yes, I included a section about that in my original technical description. it should be covered here if it's not covered later in the document. At leas a note/link here to where it can be configured later.

> Static web assets are discovered at build time and optimized using content-based [fingerprinting](https://wikipedia.org/wiki/Fingerprint_(computing)) to prevent the reuse of old file and [compressed](/aspnet/core/performance/response-compression) to reduce asset delivery time. [The optimizations don't include minification.]

Reviewer: ```suggestion
Static web assets are discovered at build time and optimized using content-based [fingerprinting](https://wikipedia.org/wiki/Fingerprint_(computing)) to prevent the reuse of old files.  Assets are also [compressed](/aspnet/core/performance/response-compression) to reduce asset delivery time.
```

At runtime, the discovered static web assets are exposed as endpoints with HTTP headers applied, such as [caching headers](https://developer.mozilla.org/docs/Web/HTTP/Headers/Cache-Control) and content type headers. An asset is served once until the file changes or the browser clears its cache. The [`ETag`](https://developer.mozilla.org/docs/Web/HTTP/Headers/ETag), [`Last-Modified`](https://developer.mozilla.org/docs/Web/HTTP/Headers/Last-Modified), and [`Content-Type`](https://developer.mozilla.org/docs/Web/HTTP/Reference/Headers/Content-Type) headers are set. The browser is prevented from using stale assets after an app is updated.

Delivery of static assets is based on [endpoint routing](xref:fundamentals/routing), so it works with other endpoint-aware features, such as authorization. It's designed to work with all UI frameworks, including Blazor, Razor Pages, and MVC.

Map Static Assets provides the following benefits:

> * Build-time compression for all the assets in the app, including JavaScript (JS) and stylesheets but excluding image and font assets that are already compressed. [Gzip](https://tools.ietf.org/html/rfc1952) (`Content-Encoding: gz`) compression is used during development. Gzip with [Brotli](https://tools.ietf.org/html/rfc7932) (`Content-Encoding: br`) compression is used during publish.

Reviewer: ```suggestion
* Build-time compression for all the assets in the app, including JavaScript (JS) and stylesheets but excluding image and font assets that are already compressed. [Gzip](https://tools.ietf.org/html/rfc1952) (`Content-Encoding: gz`) compression is used during development. Gzip and [Brotli](https://tools.ietf.org/html/rfc7932) (`Content-Encoding: br`) compression are both used during publish.
```

* Place code files, including `.cs` and `.cshtml`, outside of the app project's [web root](xref:fundamentals/index#web-root). A logical separation is therefore created between the app's client-side content and server-based code. This prevents server-side code from being leaked.

## MSBuild properties

The following tables show the static files MSBuild properties and metadata descriptions.

Property | Description
--- | ---
`EnableDefaultCompressedItems` | Enables default compression include/exclude patterns.
`CompressionIncludePatterns` | Semicolon-separated list of file patterns to include for compression.
`CompressionExcludePatterns` | Semicolon-separated list of file patterns to exclude from compression.
`EnableDefaultCompressionFormats` | Enables default compression formats (Gzip and Brotli).
`BuildCompressionFormats` | Compression formats to use during build.
`PublishCompressionFormats` | Compression formats to use during publish.
`DisableBuildCompression` | Disables compression during build.
`CompressDiscoveredAssetsDuringBuild` | Compresses discovered assets during build.
`BrotliCompressionLevel` | Compression level for the Brotli algorithm.

> `LinkAlternativeRepresentationsToOriginalResource` | How to link compressed assets to original resources.

Reviewer 2: We should remove this property, I missed it when I created the list, it should not be here.

> `LinkAlternativeRepresentationsToOriginalResource` | How to link compressed assets to original resources.

Author: ```suggestion
```

`StaticWebAssetBuildCompressAllAssets` | Compresses all assets during build, not just assets discovered or computed during a build.
`StaticWebAssetPublishCompressAllAssets` | Compresses all assets during publish, not just assets discovered or computed during a build.

Property | Description
--- | ---
`StaticWebAssetBasePath` | Base URL path for all the assets in a library.
`StaticWebAssetsFingerprintContent` | Enables content fingerprinting for cache busting.
`StaticWebAssetFingerprintingEnabled` | Enables fingerprinting feature for static web assets.
`StaticWebAssetsCacheDefineStaticWebAssetsEnabled` | Enables caching for static web asset definitions.
`StaticWebAssetBuildManifestPath` | Path to the build manifest JSON file.
`StaticWebAssetsBuildManifestCacheFilePath` | Path to the build manifest cache file.
`StaticWebAssetEndpointExclusionPattern` | Pattern for excluding endpoints.

Item group | Description | Metadata
--- | ---

> `StaticWebAssetContentTypeMapping` | Maps file patterns to content types and cache headers for endpoints. | Pattern, Cache

Reviewer 2: This should be covered under a section about "adding new content type mappings at build time"

Author: Ok ... I have one more pass to get the sample app out of here. It's in the samples repo now. I didn't do it yet because I didn't want to clutter the diff. I'll take care of it by EOD.
